gordon_janitor_gcp/gpubsub_client.py

```diff
@@ -0,0 +1,224 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Client module to publish any required DNS changes initiated from
+:py:class:`gordon_janitor_gcp.gdns_reconciler.GoogleDNSReconciler` to
+Google Cloud Pub/Sub. The consumer of these messages is the `Gordon
+service <https://github.com/spotify/gordon>`_
+
+This client wraps around `google-cloud-pubsub <https://pypi.python.org
+/pypi/google-cloud-pubsub>`_ using `grpc <https://github.com/googleapis/
+googleapis/blob/master/google/pubsub/v1/pubsub.proto>`_ rather than
+inheriting from :py:class:`gordon_janitor_gcp.http_client.\
+AIOGoogleHTTPClient`.
+
+.. attention::
+
+    This publisher client is an internal module for the core janitor
+    logic. No other use cases are expected.
+
+
+To use:
+
+.. code-block:: python
+
+    import asyncio
+
+    config = {
+        'keyfile': '/path/to/keyfile.json',
+        'project': 'a-dns-project'
+    }
+    changes_channel = asyncio.Queue()
+
+    publisher = GooglePubsubPublisher(
+        config, changes_channel)
+
+    loop = asyncio.get_event_loop()
+    try:
+        loop.run_until_complete(publisher.start())
+    finally:
+        loop.close()
+"""
+
+import asyncio
+import datetime
+import json
+import logging
+import os
+
+from asyncio_extras import threads
+from google.api_core import exceptions as google_exceptions
+from google.cloud import pubsub
+
+from gordon_janitor_gcp import auth
+from gordon_janitor_gcp import exceptions
+
+
+class GooglePubsubPublisher:
+    """Client to publish change messages to Google Pub/Sub.
+
+    Args:
+        config (dict): Google Cloud Pub/Sub-related configuration.
+        changes_channel (asyncio.Queue): queue to publish message to
+            make corrections to Cloud DNS.
+
+    Raises:
+        :py:exc:`gordon_janitor_gcp.exceptions.GCPGordonJanitorError`
+            if there are issues creating a Pub/Sub topic. This is not
+            raised if the topic already exists.
+
+        :py:exc:`gordon_janitor_gcp.exceptions.GCPConfigError`
+            when required configuration is not set. See :doc:`config`
+            for what is required.
+    """
+
+    def __init__(self, config, changes_channel=None, **kw):
+        self.config = config
+        self.changes_channel = changes_channel
+        self.cleanup_timeout = config.get('cleanup_timeout', 60)
+        self.topic = None  # set in _init
+        self.publisher = None  # set in _init
+        self._messages = set()
+        self._init()
+
+    def _init(self):
```

Comment:
Maybe I don't quite get your concerns; It's essentially mimicking current behavior of other plugins - I didn't implement a separate pubsub client because the one that google-cloud-python provides is sufficient (uses grpc + threads already). So the separation doesn't quite make sense to me - there is an API client defined elsewhere, if it's external, it doesn't make much sense to bring that into an internal module where no more additional functionality is needed. 

 I also wouldn't anticipate the pubsub API client itself to change since the google-cloud-python library explicitly says [the pubsub client library is in beta](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/README.rst) and "is being prepared for release" (which I take to assume the lib's API is stable).

Also just to note: the reconciler class does have similar two inits; if changes are made here, it will also need to be addressed there.

IMO (and again, open to be convinced otherwise) I think this plugin's responsibilities are appropriate. The janitor loads the plugin, initializes it, then uses the initialized object to start the work required to happen. I suppose one approach to address your "it's doing too much" concern is to have the plugin entry point in `setup.py` to actually point to a function that parses config, creates an auth client, creates a publisher client, and returns an initialized  Publisher plugin with what it needs, so then the janitor still receives an instantiated plugin object to call `start` on. It's six of one/half dozen of another in my PoV, so if you do have a preference for a new function (or a different approach), I'd be alright with that.